typing/parmatch.ml

```diff
@@ -44,6 +44,132 @@ let omega_list l = List.map (fun _ -> omega) l
 
 let zero = make_pat (Tpat_constant (Const_int 0)) Ctype.none Env.empty
 
+module Simple_head_pat : sig
+  type desc =
+    | Any
+    | Construct of constructor_description
+    | Constant of constant
+    | Tuple of int
+    | Record of label_description list
+    | Variant of { tag: label; has_arg: bool; row: row_desc ref }
+    | Array of int
+    | Lazy
+
+  type t
+
+  val desc : t -> desc
+  val env : t -> Env.t
+  val loc : t -> Location.t
+  val typ : t -> Types.type_expr
+
+  (** [of_pattern p] raises [Invalid_arg _] if [p] is an or-pattern or an
+      exception-pattern. *)
+  val of_pattern : pattern -> t
+
+  (** reconstructs a pattern, putting wildcards as sub-patterns. *)
+  val to_pattern : t -> pattern
+
+  val make_record
+    :  loc:Location.t
+    -> typ:Types.type_expr
+    -> env:Env.t
+    -> label_description list
+    -> t
```

Comment:
I'm not fond of this function. At first I thought that `_record` here referred to the internal implementation of `t` as a record. Couldn't we just have the more general

```ocaml
val make : loc -> typ -> env -> desc -> t
```

?

Reply:
> At first I thought that _record here referred to the internal implementation of t as a record

That would have been a pretty bad name.

---

We could have the more general function sure, `make_record` does feel weird.